crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,567 @@ impl State {
+            // value and record the share proceeds. We'll ultimately return the
+            // share proceeds that resulted in the smallest value.
+            let delta = I256::try_from(present_value.mul_down(lp_total_supply))?
+                - I256::try_from(starting_present_value.mul_up(active_lp_total_supply))?;
+            if smallest_delta == I256::from(0) || delta.abs() < smallest_delta.abs() {
+                smallest_delta = delta;
+                closest_share_proceeds = share_proceeds;
+                closest_present_value = present_value;
+            }
+
+            // We calculate the updated share proceeds `x_n+1` by proceeding
+            // with Newton's method. This is given by:
+            //
+            // x_n+1 = x_n - F(x_n) / F'(x_n)
+            if delta > I256::zero() {
+                // NOTE: Round the quotient down to avoid overshooting.
+                share_proceeds = share_proceeds
+                    + FixedPoint::from(delta)
+                        .div_down(derivative)
+                        .div_down(lp_total_supply);
+            } else if delta < I256::zero() {
+                let delta = FixedPoint::from(-delta)
+                    .div_down(derivative)
+                    .div_down(lp_total_supply);
+                if delta < share_proceeds {
+                    share_proceeds = share_proceeds - delta;
+                } else {
+                    return Err(eyre!("Delta is greater than share proceeds."));
+                }
+            } else {
+                break;
+            }
+        }
+
+        let updated_state =
+            self.get_state_after_liquidity_update(-I256::try_from(share_proceeds)?)?;
+        let present_value = updated_state.calculate_present_value(current_block_timestamp)?;
+
+        // Check to see if the current share proceeds is closer to the optimal
+        // value than the previous closest value. We'll choose whichever of the
+        // share proceeds that is closer to the optimal value.
+        let last_delta = I256::try_from(present_value.mul_down(lp_total_supply))?
+            - I256::try_from(starting_present_value.mul_up(active_lp_total_supply))?;
+        if last_delta.abs() < smallest_delta.abs() {
+            closest_share_proceeds = share_proceeds;
+            closest_present_value = present_value;
+        }
+
+        // Verify that the LP share price was conserved within a reasonable
+        // tolerance.
+        if closest_present_value.div_down(active_lp_total_supply)
+            < starting_present_value.mul_div_up(
+                fixed!(1e18) - FixedPoint::from(SHARE_PROCEEDS_TOLERANCE),
+                lp_total_supply,
+            )
+            || closest_present_value.div_up(active_lp_total_supply)
+                > starting_present_value.mul_div_down(
+                    fixed!(1e18) + FixedPoint::from(SHARE_PROCEEDS_TOLERANCE),
+                    lp_total_supply,
+                )
```

Comment:
This is what the solidity looks like:

```
        if (
            // NOTE: Round down to make the check stricter.
            closestPresentValue.divDown(params.activeLpTotalSupply) <
            params.startingPresentValue.mulDivUp(
                ONE - SHARE_PROCEEDS_TOLERANCE,
                lpTotalSupply
            ) ||
            // NOTE: Round up to make the check stricter.
            closestPresentValue.divUp(params.activeLpTotalSupply) >
            params.startingPresentValue.mulDivDown(
                ONE + SHARE_PROCEEDS_TOLERANCE,
                lpTotalSupply
            )
        ) {
            // NOTE: Return 0 to indicate that the share proceeds couldn't be
            // calculated.
            return 0;
        }
```

Reply:
Ok, made the rust match this as close as prettier lets me